### DISCUSSION TITLE
Exiting fzf with Chafa image preview doesn't clear Kitty images

### DISCUSSION THREAD

**Discussion by onaforeignshore** [ghostty-org/ghostty#3054] [Q&A]
**Steps to reproduce:**

1. Install https://github.com/junegunn/fzf (I used `brew install fzf`).
2. Configure FZF to use the supplied `fzf-preview.sh` file.
```zsh
export FZF_CTRL_T_OPTS="--preview '/opt/homebrew/bin/fzf-preview.sh {}'"
```
3. Open the terminal in a folder with images and press `Ctrl-T`.
4. Press `Esc` to exit FZF.

**Expected output:**
After exiting, the terminal background clears.

https://github.com/user-attachments/assets/e1e65d54-9478-4419-8819-501db775b2d4

_Example output in iTerm_

**Actual output:**
After exiting, the image remains until you do a `clear` command. Subsequent terminal commands have the text hidden behind the image.

https://github.com/user-attachments/assets/40a98f3c-82ec-4e1b-a84b-c4902664877c

_Example output in ghostty_

**Configuration:**
```
macOS Sequoia 15.2 (24C101)

❯ ghostty --version
Ghostty 0.1.0-main+05fd5e0c

Version
  - version: 0.1.0-main+05fd5e0c
  - channel: tip
Build Config
  - Zig version: 0.13.0
  - build mode : builtin.OptimizeMode.ReleaseFast
  - app runtime: apprt.Runtime.none
  - font engine: font.main.Backend.coretext
  - renderer   : renderer.Metal
  - libxev     : main.Backend.kqueue
```

**Comment by mitchellh** [comment]
@rockorager Do you happen to know about this one? If not, no worries just wanted to quickly check.

**Comment by rockorager** [comment]
I looked at the script and they use a different command on kitty. I'd be curious if this happens on wezterm also. But we'd need to see if it's using sixel or kitty on wezterm. It could be an issue with the script never deleting the image, and until ghostty there has never been another terminal implementing kitty graphics but not sixels.

**Reply by mitchellh** [reply]
Got it. I'm going to convert to Q&A for now then because this _might_ be an upstream issue. If we determine its a VT issue I'll spin out a focused bug issue for that for us.

**Reply by JohnFinn** [reply]
I have a similar setup with `fzf & chafa` on nixos. On wezterm it works as expected. This prevents me from switching to ghostty.
```fish
function fzf --wraps="FZF_DEFAULT_COMMAND='/nix/store/3j75iyzm2fyh03m21ix5296h8hd06q8z-fd-10.2.0/bin/fd --type f' /nix/store/z3ayhjslz72ldiwrv3mn5n7rs96p2g8a-fzf-0.62.0/bin/fzf --preview '/nix/store/fnshacq1q21mgsripyab83888p99mkvg-chafa-1.14.5-bin/bin/chafa --view-size \${FZF_PREVIEW_COLUMNS}x\${FZF_PREVIEW_LINES} {} 2> /dev/null || /nix/store/119y38i56azh5iflcwwwkws52h40y60d-bat-0.25.0/bin/bat --color=always {}' --bind ctrl-k:down,ctrl-l:up" --description "alias fzf FZF_DEFAULT_COMMAND='/nix/store/3j75iyzm2fyh03m21ix5296h8hd06q8z-fd-10.2.0/bin/fd --type f' /nix/store/z3ayhjslz72ldiwrv3mn5n7rs96p2g8a-fzf-0.62.0/bin/fzf --preview '/nix/store/fnshacq1q21mgsripyab83888p99mkvg-chafa-1.14.5-bin/bin/chafa --view-size \${FZF_PREVIEW_COLUMNS}x\${FZF_PREVIEW_LINES} {} 2> /dev/null || /nix/store/119y38i56azh5iflcwwwkws52h40y60d-bat-0.25.0/bin/bat --color=always {}' --bind ctrl-k:down,ctrl-l:up"
  FZF_DEFAULT_COMMAND='/nix/store/3j75iyzm2fyh03m21ix5296h8hd06q8z-fd-10.2.0/bin/fd --type f' /nix/store/z3ayhjslz72ldiwrv3mn5n7rs96p2g8a-fzf-0.62.0/bin/fzf --preview '/nix/store/fnshacq1q21mgsripyab83888p99mkvg-chafa-1.14.5-bin/bin/chafa --view-size ${FZF_PREVIEW_COLUMNS}x${FZF_PREVIEW_LINES} {} 2> /dev/null || /nix/store/119y38i56azh5iflcwwwkws52h40y60d-bat-0.25.0/bin/bat --color=always {}' --bind ctrl-k:down,ctrl-l:up $argv

end
```
https://github.com/user-attachments/assets/e1f7fda9-b140-4211-bc64-c6d5efe0bdb5

Also when passing `-f iterm` to `chafa`, `wezterm` keeps working, whereas `ghostty` stops showing any images.

**Reply by rhodes-b** [reply]
@JohnFinn 

Ghostty does not support iterm image protocol it never claimed too (wezterm does!)

wezterm does not work with kitty images with fzf + chafa (even though just chafa -f kitty seems to work fine)

as such its using the iterm protocol and does so by default
https://github.com/hpjansson/chafa/blob/9ac25a52237f2bf5fbb82a9a3f7fbe3e7007f470/chafa/chafa-term-db.c#L787-L794

I with that exact same command you just sent it works just fine for me ghostty 1.1.3

https://github.com/user-attachments/assets/869b91d5-5910-4e83-8f5b-3c831f2c4f9d





I also dont see the same issue as op does in 1.1.3 with the preview script they both work as expected in ghostty using kitten icat
 or chafa (post right above this)

kitten icat has the benefit of not overlapping images which chafa does this also happens in kitty when using chafa so no its not a ghostty bug

https://github.com/junegunn/fzf/blob/master/bin/fzf-preview.sh

**Reply by JohnFinn** [reply]
> I with that exact same command you just sent it works just fine for me ghostty 1.1.3

@rhodes-b your scenario does not fully reproduce the issue. There seems to be a single image in the directory. The issue reproduces when there are images of different sizes and text files, and you're switching between them

**Reply by rhodes-b** [reply]
I tried that as well and had no issue. Like stated images do have overlapping using chafa but do not using kitten icat

Kitty has the same behavior so that is not a ghostty bug 

I only showed a video with a single image because I didnt want to show more information than that (ie my homedir which is what i tested)

**Comment by infused-kim** [comment]
I am experiencing the same issue with fzf and [chafa](https://github.com/hpjansson/chafa) v1.14.5 (latest) as the preview tool on macOS Sonoma 14.5.

Chafa has a `-f` option to select the format:
```shell
  -f, --format=FORMAT  Set output format; one of [iterm, kitty, sixels,
                     symbols]. Iterm, kitty and sixels yield much higher
                     quality but enjoy limited support. Symbols mode yields
                     beautiful character art.
```

But only `kitty` and `symbols` output anything. The rest show no image, but also no errors.

The `-f symbols` option clears the previous image when I switch between fzf results, but is of course of very low quality.

**Reply by rhodes-b** [reply]
Ghostty doesnt support iterm image protocol or sixels, so those wont work. It would be expected that kitty and symbols work as symbols just uses unicode / ascii characters with color codes and ghostty supports kitty image protocol
